File: 0046-permutations/0046-permutations.go
func permute(nums []int) [][]int {
    res:=[][]int{}
    if len(nums)==1{
        return append(res,append([]int{},nums...))
    }
    var dfs func(start,end int)
    dfs=func(start,end int){
        if start==len(nums){
            res=append(res,append([]int{},nums...))
            return
        }
        for i:=start;i<end;i++{
            nums[start],nums[i]=nums[i],nums[start]
            dfs(start+1,end)
            nums[start],nums[i]=nums[i],nums[start]
        }
    }
    dfs(0,len(nums))
    return res
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums)==1:
            return [nums]
        res=[]
        path=[]
        used=[False]*len(nums)
        def dfs():
            if len(path)==len(nums):
                res.append(path[:])
                return 
            for i in range(len(nums)):
                if used[i]:
                    continue
                path.append(nums[i])
                used[i]=True
                dfs()
                used[i]=False
                path.pop()
        dfs()
        return res